pkg/inventory/inventory-client.go

```diff
@@ -6,473 +6,402 @@ package inventory
+		inventories = append(inventories, uInv)
 	}
-
-	return identifiers, nil
+	return inventories, nil
 }
 
-// deleteInventoryObjByName deletes the passed inventory object from the APIServer, or
-// an error if one occurs.
-func (cic *ClusterClient) deleteInventoryObjByName(ctx context.Context, obj *unstructured.Unstructured, dryRun common.DryRunStrategy) error {
-	if obj == nil {
-		return fmt.Errorf("attempting delete a nil inventory object")
-	}
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run delete inventory object: not deleted")
+// CreateOrUpdate the in-cluster inventory
+// Updates the unstructured object, or creates it if it doesn't exist
+func (cic *UnstructuredClient) CreateOrUpdate(ctx context.Context, inv Inventory, _ UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
+	}
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
+	}
+	// Attempt to retry on a resource conflict error to avoid needing to retry the
+	// entire Apply/Destroy when there's a transient conflict.
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		create := false
+		obj, err := cic.client.Namespace(ui.Namespace()).Get(ctx, ui.Name(), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			create = true
+		} else if err != nil {
+			return err
+		}
+		if obj != nil { // Obj is nil when IsNotFound, in this case keep initial/empty obj
+			ui.ClusterObj = obj
+		}
+		uObj, err := cic.toUnstructured(ui)
+		if err != nil {
+			return err
+		}
+		var newObj *unstructured.Unstructured
+		if create {
+			klog.V(4).Infof("creating inventory object %s/%s/%s", cic.gvk, uObj.GetNamespace(), uObj.GetName())
+			newObj, err = cic.client.Namespace(uObj.GetNamespace()).Create(ctx, uObj, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			klog.V(4).Infof("updating inventory object %s/%s/%s", cic.gvk, uObj.GetNamespace(), uObj.GetName())
+			newObj, err = cic.client.Namespace(uObj.GetNamespace()).Update(ctx, uObj, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		ui.ClusterObj = newObj
 		return nil
-	}
-
-	mapping, err := cic.getMapping(obj)
+	})
 	if err != nil {
 		return err
 	}
-
-	klog.V(4).Infof("deleting inventory object: %s/%s", obj.GetNamespace(), obj.GetName())
-	return cic.dc.Resource(mapping.Resource).Namespace(obj.GetNamespace()).
-		Delete(ctx, obj.GetName(), metav1.DeleteOptions{})
-}
-
-// getMapping returns the RESTMapping for the provided resource.
-func (cic *ClusterClient) getMapping(obj *unstructured.Unstructured) (*meta.RESTMapping, error) {
-	return cic.mapper.RESTMapping(obj.GroupVersionKind().GroupKind(), obj.GroupVersionKind().Version)
+	if cic.statusPolicy != StatusPolicyAll {
+		return nil
+	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		obj, err := cic.client.Namespace(ui.Namespace()).Get(ctx, ui.Name(), metav1.GetOptions{})
```

Comment:
I don't know what the clean way to do this would be, but there's an opportunity to avoid a GET here, since the Create & Update calls above will have updated the uObj with the latest state from the cluster, including the status. So we don't need to GET again on the first attempt.